### PR TITLE
msvs: Support ARM64 configuration platform

### PR DESCRIPTION
**Pull request by kaadam**
* This change is intent to support top-level ARM64 configuration, so a
project can be built as ARM64 project.
* Unit test also included

### CODE REVIEW

**Comment by kaadam** [comment]
@refack I have uploaded the GYP part of this change: https://github.com/nodejs/node-gyp/pull/1678

I added a new unit test case in configurations directory for testing arm64 configuration platform, is it okay, or should I place it somewhere else. Should it be more complex? Since I copied this test case from 'x64', I just did minor changes on that, I prevent the licence as it was in 'x64'.

**Comment by refack** [comment]
Hi @kaadam and thank you for the contribution.
This needs a rebase (I'd be happy to do it if you check the `Allow edits from maintainers.` box)
image

**Comment by kaadam** [comment]
@refack Hi, This is weird as far as I see that box was ticked. But I have rebased the change.

**Comment by kaadam** [comment]
@refack could you look at this change?

File: test/configurations/arm64/gyptest-arm64.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import sys
+
+formats = ['msvs']
+if sys.platform == 'win32':
+  formats += ['ninja']
+test = TestGyp.TestGyp(formats=formats)
+
+test.run_gyp('configurations.gyp')
+test.set_configuration('Debug|ARM64')
```
Comment (refack) [inline test/configurations/arm64/gyptest-arm64.py]: On VS2017 I'm getting `The specified solution configuration "Debug|ARM64|x64" is invalid.`

Comment (kaadam) [reply test/configurations/arm64/gyptest-arm64.py]: okay, could you send me how I can reproduce this?

File: gyp/MSVS/MSVSSettings.py [hunk 2]
```diff
@@ -978,6 +978,7 @@ def _ValidateSettings(validators, settings, stderr):
 _Same(_midl, 'TargetEnvironment',
       _Enumeration(['NotSet',
                     'Win32',  # /env win32
+                    'ARM64',  # /env arm64
```
[the remark this answers is not in the document]
Comment (kaadam) [reply gyp/MSVS/MSVSSettings.py]: Yes, it seems it is not relevant to this change. I do a double-check.

[the remark this answers is not in the document]
Comment (jkunkee) [reply gyp/MSVS/MSVSSettings.py]: This is related to MIDL, a code-generation tool in the MSVC build system. You would need to have a MIDL target in your GYP file to test this; I'm not sure how to do that. I believe MIDL does need the flag.

If anything, moving this to the end of the array is what @refack is requesting.

File: test/configurations/arm64/gyptest-arm64.py (hunk 1 above)
Comment (jkunkee) [reply test/configurations/arm64/gyptest-arm64.py]: Note that VS2019 isn't out of beta yet, so it's possible that this is a VS bug.

Comment (refack) [reply test/configurations/arm64/gyptest-arm64.py]:
```cmd
python gyptest.py -v -f msvs test/configurations/arm64/gyptest-arm64.py
```

Comment (refack) [reply test/configurations/arm64/gyptest-arm64.py]: I think it's a regression I introduced. Chasing this down.

File: gyp/MSVS/MSVSSettings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (refack) [reply gyp/MSVS/MSVSSettings.py]: Fixed

File: gyp/msvs_emulation.py [hunk 3]
```diff
@@ -921,7 +933,7 @@ def GenerateEnvironmentFiles(toplevel_build_dir, generator_flags):
   if generator_flags.get('ninja_use_custom_environment_files', False) or not vs.path:
     return
 
-  archs = ('x86', 'x64')
+  archs = ('x86', 'x64', 'arm64')
```
[the remark this answers is not in the document]
Comment (refack) [reply gyp/msvs_emulation.py]: I reverted this for the time being

File: test/configurations/arm64/gyptest-arm64.py (hunk 1 above)
Comment (refack) [reply test/configurations/arm64/gyptest-arm64.py]: Solved.

**Comment by refack** [comment]
I solved the regressing with the MSBuild triggering. The new test passes.

**Comment by kaadam** [comment]
@refack Thanks for improving this and your help. So Shall I do something with it? Could it be merged?

**Comment by refack** [comment]
> Thanks for improving this and your help. So Shall I do something with it? Could it be merged?

I just need to get the "lint" out and I'll land this.

**Pull request merged**